lib/Dialect/Torch/Transforms/RefineTypes.cpp

```diff
@@ -1176,26 +1176,27 @@ ChangeResult TypeAnalyzer::visitAtenNllLossForwardOp(
   auto self = operands[0]->getValue();
   auto outputKnowledge =
       ValueKnowledge::getNotNonePessimisticValueState(op.getContext());
-
-  // Contains Knowledge of shape and dtype for the 1st result.
-  outputKnowledge.dtype = self.dtype;
-  int64_t reduction;
-  unsigned resultRank = self.sizes.size();
-
-  // Contains Knowledge of shape and dtype for the 2nd result.
   auto totalWeightKnowledge =
       ValueKnowledge::getNotNonePessimisticValueState(op.getContext());
+
+  // Contains Knowledge of dtype for both results.
+  outputKnowledge.dtype = self.dtype;
   totalWeightKnowledge.dtype = self.dtype;
-  totalWeightKnowledge.sizes.resize(0, kUnknownSize);
-  totalWeightKnowledge.hasSizes = true;
+  if (self.hasSizes) {
+    int64_t reduction;
+    unsigned resultRank = self.sizes.size();
 
-  if (self.hasSizes &&
-      matchPattern(op.reduction(), m_TorchConstantInt(&reduction))) {
-    if (reduction != Reduction::None)
-      resultRank -= 1;
+    // Contains Knowledge of shape for the both results.
+    // totalWeight is a 0d tensor.
+    totalWeightKnowledge.hasSizes = true;
+
+    if (matchPattern(op.reduction(), m_TorchConstantInt(&reduction))) {
+      if (reduction != Reduction::None)
+        resultRank -= 1;
```

Comment:
I think I had made a comment about this last time. I still don't understand why we subtract one here. From the documentation, (https://pytorch.org/docs/stable/generated/torch.nn.NLLLoss.html#torch.nn.NLLLoss), the input can be a rank-1 or rank-2 tensor. If the tensor is rank-1, and `reduction != None`, then `outputKnowledge.sizes` will be resized to size `-1`, which is not right. If the reduction is not `None` the result should be a scalar, so `resultRank = 0`. 

I would flip the condition in the `if` and do this instead

```cpp
outputKnowledge.hasSizes = true;
if (reduction == Reduction::None) {
  outputKnowledge.sizes.resize(resultRank - 1, kUnknownSize);
}
```

leaving `outputKnowledge.sizes` empty represents the scalar case.

Reply:
I think I missed this last time. Thanks for pointing this out and Sorry for the re-review!